pkg/ko/publish/daemon.go

```diff
@@ -49,26 +49,27 @@ func (d *demon) Publish(img v1.Image, s string) (name.Reference, error) {
 		return nil, err
 	}
 
-	tagsAndDigest := append(d.tags, h.Hex)
-
-	for _, tagName := range tagsAndDigest {
+	var tags []name.Tag
+	for _, tagName := range d.tags {
 		tag, err := name.NewTag(fmt.Sprintf("%s/%s:%s", LocalDomain, d.namer(s), tagName), name.WeakValidation)
 		if err != nil {
 			return nil, err
 		}
-		log.Printf("Loading %v", tag)
-		// TODO: This is slow because we have to load the image multiple times.
-		// We should use client.ImageTag instead
-		if _, err := daemon.Write(tag, img); err != nil {
-			return nil, err
-		}
-		log.Printf("Loaded %v", tag)
+
+		tags = append(tags, tag)
 	}
 
-	tag, err := name.NewTag(fmt.Sprintf("%s/%s:%s", LocalDomain, d.namer(s), h.Hex), name.WeakValidation)
+	digestTag, err := name.NewTag(fmt.Sprintf("%s/%s:%s", LocalDomain, d.namer(s), h.Hex), name.WeakValidation)
 	if err != nil {
 		return nil, err
 	}
+	tags = append(tags, digestTag)
+
+	log.Printf("Loading %v", tags)
+	if _, err := daemon.WriteAndTag(tags, img); err != nil {
```

Comment:
I'd call `daemon.Write` with the `digestTag`, then loop over `d.tags` and call `daemon.Tag` for each of them. Then we can keep the `daemon` API simpler.

Reply:
Indeed, I personally liked the result a lot, way simpler and more readable, thanks for the suggestion  :smile: